File: kotlin/src/main/kotlin/RotationalCipher.kt
class RotationalCipher(val offset: Int) {
    fun encode(text: String): String {
        return String(text.map {
            if (it.isLetter()) {
                var raw = it + offset
                if (it.isUpperCase() && raw > 'Z') raw -= 26
                if (it.isLowerCase() && raw > 'z') raw -= 26
                raw
            } else {
                it
            }
        }.toCharArray())
    }
}

File: scala/src/main/scala/RotationalCipher.scala
object RotationalCipher {
  private def rotateLetter(c: Char, offset: Int) = {
    val baseLetter = if (c.isLower) 'a' else 'A'
    ((c - baseLetter + offset) % 26 + baseLetter).asInstanceOf[Char].toString
  }

  def rotate(str: String, offset: Int): String =
    str.map(c => if (c.isLetter) rotateLetter(c, offset % 26) else c.toString).mkString
}